pkg/scheduler/algorithm/predicates/metadata.go

```diff
@@ -186,61 +177,90 @@ func (m *serviceAffinityMetadata) clone() *serviceAffinityMetadata {
 }
 
 type podAffinityMetadata struct {
-	topologyPairsAntiAffinityPodsMap *topologyPairsMaps
-	// A map of topology pairs to a list of Pods that can potentially match
-	// the affinity terms of the "pod" and its inverse.
-	topologyPairsPotentialAffinityPods *topologyPairsMaps
-	// A map of topology pairs to a list of Pods that can potentially match
-	// the anti-affinity terms of the "pod" and its inverse.
-	topologyPairsPotentialAntiAffinityPods *topologyPairsMaps
-}
-
-func (m *podAffinityMetadata) addPod(addedPod *v1.Pod, pod *v1.Pod, node *v1.Node) error {
-	// Add matching anti-affinity terms of the addedPod to the map.
-	topologyPairsMaps, err := getMatchingAntiAffinityTopologyPairsOfPod(pod, addedPod, node)
-	if err != nil {
-		return err
-	}
-	m.topologyPairsAntiAffinityPodsMap.appendMaps(topologyPairsMaps)
-	// Add the pod to nodeNameToMatchingAffinityPods and nodeNameToMatchingAntiAffinityPods if needed.
-	affinity := pod.Spec.Affinity
-	podNodeName := addedPod.Spec.NodeName
-	if affinity != nil && len(podNodeName) > 0 {
-		// It is assumed that when the added pod matches affinity of the pod, all the terms must match,
-		// this should be changed when the implementation of targetPodMatchesAffinityOfPod/podMatchesAffinityTermProperties
-		// is changed
-		if targetPodMatchesAffinityOfPod(pod, addedPod) {
-			affinityTerms := GetPodAffinityTerms(affinity.PodAffinity)
-			for _, term := range affinityTerms {
-				if topologyValue, ok := node.Labels[term.TopologyKey]; ok {
-					pair := topologyPair{key: term.TopologyKey, value: topologyValue}
-					m.topologyPairsPotentialAffinityPods.addTopologyPair(pair, addedPod)
+	// A map of topology pairs to the number of existing pods that has anti-affinity terms that match the "pod".
+	topologyToMatchedExistingAntiAffinityTerms topologyToMatchedTermCount
+	// A map of topology pairs to the number of existing pods that match the affinity terms of the "pod".w
+	topologyToMatchedAffinityTerms topologyToMatchedTermCount
+	// A map of topology pairs to the number of existing pods that match the anti-affinity terms of the "pod".
+	topologyToMatchedAntiAffinityTerms topologyToMatchedTermCount
+}
+
+// updateWithAffinityTerms updates the topologyToMatchedTermCount map with the specified value
+// for each affinity term if "targetPod" matches ALL terms.
+func (m topologyToMatchedTermCount) updateWithAffinityTerms(targetPod *v1.Pod, targetPodNode *v1.Node, affinityTerms []*affinityTermProperties, value int64) {
+	if podMatchesAllAffinityTermProperties(targetPod, affinityTerms) {
+		for _, t := range affinityTerms {
+			if topologyValue, ok := targetPodNode.Labels[t.term.TopologyKey]; ok {
+				pair := topologyPair{key: t.term.TopologyKey, value: topologyValue}
+				m[pair] += value
+				// value could be a negative value, hence we delete the entry if
+				// the entry is down to zero.
+				if m[pair] == 0 {
```

Comment:
Same here: checking `if m[pair] == 0` works for both a non-existing entry or existing entry but with value 0.

- if we want to manually delete the entry, we'd better keep the logic consistent to use `if _, ok := m[pair]; ok` on the logic of checking its existence
- if we don't delete the entry, probably it increases some memory footprint, but we save time on checking and deletion. in this case, checking `if m[pair] == 0` should be used to check its existence.

Reply:
since the previous logic relied on the existence of the entry rather than whether or not it is zero, I opted to do this to avoid potential bugs. Doing this also made it easy to pass the tests that does DeepEqual test. We can clean this up in a followup PR that converts the type to *int64 so that we can do atomic adds instead of using a global mutex.